packages/react/src/components/AccountSettings/ChangePassword/ChangePassword.tsx

```diff
@@ -1,53 +1,165 @@
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ): string[] => {
+      const { newPassword } = formValues;
+
+      return runFieldValidators({
+        value: newPassword,
+        validators: passwordValidators,
+        eventType,
+        hasBlurred: blurredFields.includes('newPassword'),
+      });
+    },
+    [passwordValidators]
+  );
+
+  const validateConfirmPassword = React.useCallback(
+    (
+      formValues: FormValues,
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ): string[] => {
+      const { newPassword, confirmPassword } = formValues;
+
+      const confirmPasswordValidators =
+        getDefaultConfirmPasswordValidators(newPassword);
+
+      return runFieldValidators({
+        value: confirmPassword,
+        validators: confirmPasswordValidators,
+        eventType,
+        hasBlurred: blurredFields.includes('confirmPassword'),
+      });
+    },
+    []
+  );
+
+  const runValidation = React.useCallback(
+    (
+      formValues: FormValues,
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ) => {
+      const passwordErrors = validateNewPassword(
+        formValues,
+        blurredFields,
+        eventType
+      );
+      const confirmPasswordErrors = validateConfirmPassword(
+        formValues,
+        blurredFields,
+        eventType
+      );
+
+      const newValidationError = {
+        newPassword: passwordErrors,
+        confirmPassword: confirmPasswordErrors,
+      };
 
-  /** Translations */
+      // only re-render if errors have changed
+      if (!isEqual(validationError, newValidationError)) {
+        setValidationError(newValidationError);
+      }
+    },
+    [validateConfirmPassword, validateNewPassword, validationError]
+  );
+
+  /* Translations */
   // TODO: add AccountSettingsTextUtil to collect these strings
   const currentPasswordLabel = translate('Current Password');
   const newPasswordLabel = translate('New Password');
+  const confirmPasswordLabel = translate('Confirm Password');
   const updatePasswordText = translate('Update password');
 
-  /** Event Handlers */
+  /* Event Handlers */
   const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
     event.preventDefault();
 
     const { name, value } = event.target;
-    setFormValues({ ...formValues, [name]: value });
+
+    const newFormValues = { ...formValues, [name]: value };
+    runValidation(newFormValues, blurredFields, 'change');
+
+    setFormValues(newFormValues);
+  };
+
+  const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
+    event.preventDefault();
+
+    const { name } = event.target;
+    // only update state and run validation if this is the first time blurring the field
+    if (!blurredFields.includes(name)) {
+      const newBlurredFields = [...blurredFields, name];
+      runValidation(formValues, newBlurredFields, 'blur');
+      setBlurredFields(newBlurredFields);
```

Comment:
The more i've looked at this PR, have started thinking you could probably use a ref for `blurredFields` to avoid the re-renders that triggered by `setBlurredFields`

Reply:
I agree, `blurredFields` by itself doesn't need a re-render; only blur validation does.

[39a189c](https://github.com/aws-amplify/amplify-ui/pull/2900/commits/39a189cffc3c057e461d197f3da8e5aac6954a0a)